state/application.go

```diff
@@ -350,16 +361,39 @@ func (op *DestroyApplicationOperation) destroyOps() ([]txn.Op, error) {
 	}
 	ops = append(ops, resOps...)
 
-	// We can't delete an application if it is being offered.
+	// We can't delete an application if it is being offered,
+	// unless those offers have no relations.
 	if !op.RemoveOffers {
 		countOp, n, err := countApplicationOffersRefOp(op.app.st, op.app.Name())
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
-		if n != 0 {
-			return nil, errors.Errorf("application is used by %d offer%s", n, plural(n))
+		if n == 0 {
+			ops = append(ops, countOp)
+		} else {
+			connected, err := applicationHasConnectedOffers(op.app.st, op.app.Name())
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if connected {
+				return nil, errors.Errorf("application is used by %d offer%s", n, plural(n))
+			}
+			// None of our offers are connected,
+			// it's safe to remove them.
+			removeOfferOps, err := removeApplicationOffersOps(op.app.st, op.app.Name())
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			ops = append(ops, removeOfferOps...)
+			ops = append(ops, txn.Op{
+				C:  applicationsC,
+				Id: op.app.doc.DocID,
+				Assert: bson.D{
+					{"relationcount", op.app.doc.RelationCount},
```

Comment:
Don't need both of these surely?
The the relation count changes, so too with the txn revno